File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/fibonacci.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult, require_non_negative


def memoized(n: int) -> DPResult:
    require_non_negative(n, "n")

    @lru_cache(maxsize=None)
    def solve(k: int) -> int:
        return k if k <= 1 else solve(k - 1) + solve(k - 2)

    return DPResult("Fibonacci", "memoization", solve(n))


def tabulated(n: int) -> DPResult:
    require_non_negative(n, "n")
    if n <= 1:
        return DPResult("Fibonacci", "tabulation", n, {"sequence": list(range(n + 1))})
    dp = [0] * (n + 1)
    dp[1] = 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return DPResult("Fibonacci", "tabulation", dp[n], {"sequence": dp})


def space_optimized(n: int) -> DPResult:
    require_non_negative(n, "n")
    previous, current = 0, 1
    for _ in range(n):
        previous, current = current, previous + current
    return DPResult("Fibonacci", "space optimized", previous)

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/Fibonacci.java
package com.example.dp;

import java.util.HashMap;
import java.util.Map;

public final class Fibonacci {
    private Fibonacci() {
    }

    public static Result memoized(int n) {
        return new Result("Fibonacci", "memoization", solve(n, new HashMap<>()), Map.of());
    }

    private static int solve(int n, Map<Integer, Integer> memo) {
        if (n <= 1) {
            return n;
        }
        if (memo.containsKey(n)) {
            return memo.get(n);
        }
        int value = solve(n - 1, memo) + solve(n - 2, memo);
        memo.put(n, value);
        return value;
    }

    public static Result tabulated(int n) {
        if (n <= 1) {
            return new Result("Fibonacci", "tabulation", n, Map.of());
        }
        int[] dp = new int[n + 1];
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return new Result("Fibonacci", "tabulation", dp[n], Map.of("sequence_length", dp.length));
    }

    public static Result spaceOptimized(int n) {
        int previous = 0;
        int current = 1;
        for (int i = 0; i < n; i++) {
            int next = previous + current;
            previous = current;
            current = next;
        }
        return new Result("Fibonacci", "space optimized", previous, Map.of());
    }
}
